src/java.base/share/classes/jdk/internal/math/ToDecimal.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2003, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.math;
+
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+
+import static java.lang.Math.multiplyHigh;
+
+sealed class ToDecimal permits DoubleToDecimal, FloatToDecimal {
+    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+    /* Used for left-to-tight digit extraction */
+    static final int MASK_28 = (1 << 28) - 1;
+
+    static final int NON_SPECIAL = 0;
+    static final int PLUS_ZERO   = 0x100;
+    static final int MINUS_ZERO  = 0x200;
+    static final int PLUS_INF    = 0x300;
+    static final int MINUS_INF   = 0x400;
+    static final int NAN         = 0x500;
```

Comment:
These constants could be like the original ones (0, 1, 2, ...).
This would make the switch over them a `tableswitch` rather than a `lookupswitch` in bytecode. I'm not sure whether this would improve machine code, however.
Of course, logic would need to be changed in some places (`<< 8`, `>>> 8`, etc.)